txnsync/peer.go

```diff
@@ -0,0 +1,809 @@
+	// need to update the lastTransactionSelectionGroupCounter since it's already ahead of everything in the pending transactions.
+	if grpIdx >= 0 && startIndex < len(pendingTransactions) {
+		if grpIdx == len(pendingTransactions) {
+			if grpIdx > 0 {
+				p.lastTransactionSelectionTracker.set(p.requestedTransactionsOffset, p.requestedTransactionsModulator, pendingTransactions[grpIdx-1].GroupCounter+1)
+			}
+		} else {
+			p.lastTransactionSelectionTracker.set(p.requestedTransactionsOffset, p.requestedTransactionsModulator, pendingTransactions[grpIdx].GroupCounter)
+		}
+	}
+
+	if !hasMorePendingTransactions {
+		// we're done with the current sequence.
+		p.messageSeriesPendingTransactions = nil
+	}
+
+	//fmt.Printf("selectPendingTransactions : selected %d transactions, %d not needed and aborted after exceeding data length %d/%d more = %v\n", len(selectedTxnIDs), removedTxn, accumulatedSize, windowLengthBytes, hasMorePendingTransactions)
+
+	return selectedTxns, selectedTxnIDs, hasMorePendingTransactions
+}
+
+// getLocalRequestParams returns the local requests params
+func (p *Peer) getLocalRequestParams() (offset, modulator byte) {
+	return p.localTransactionsBaseOffset, p.localTransactionsModulator
+}
+
+// update the peer once the message was sent successfully.
+func (p *Peer) updateMessageSent(txMsg *transactionBlockMessage, selectedTxnIDs []transactions.Txid, timestamp time.Duration, sequenceNumber uint64, messageSize int, filter bloomFilter) {
+	p.recentSentTransactions.addSlice(selectedTxnIDs, sequenceNumber, timestamp)
+	p.lastSentMessageSequenceNumber = sequenceNumber
+	p.lastSentMessageRound = txMsg.Round
+	p.lastSentMessageTimestamp = timestamp
+	p.lastSentMessageSize = messageSize
+	if filter.filter != nil {
+		p.lastSentBloomFilter = filter
+	}
+}
+
+// setLocalRequestParams stores the peer request params.
+func (p *Peer) setLocalRequestParams(offset, modulator uint64) {
+	if modulator > 255 {
+		modulator = 255
+	}
+	p.localTransactionsModulator = byte(modulator)
+	if modulator != 0 {
+		p.localTransactionsBaseOffset = byte(offset % modulator)
+	}
+}
+
+// peers array functions
+
+// incomingPeersOnly scan the input peers array and return a subset of the peers that are incoming peers.
+func incomingPeersOnly(peers []*Peer) (incomingPeers []*Peer) {
+	incomingPeers = make([]*Peer, 0, len(peers))
+	for _, peer := range peers {
+		if !peer.isOutgoing {
+			incomingPeers = append(incomingPeers, peer)
+		}
+	}
+	return
+}
+
+// incoming related functions
+
+func (p *Peer) addIncomingBloomFilter(round basics.Round, incomingFilter bloomFilter, currentRound basics.Round) {
+	bf := incomingBloomFilter{
+		round:  round,
+		filter: incomingFilter,
+	}
+	// scan the current list and find if we can removed entries.
+	firstValidEntry := sort.Search(len(p.recentIncomingBloomFilters), func(i int) bool {
+		return p.recentIncomingBloomFilters[i].round >= currentRound.SubSaturate(1)
+	})
+	if firstValidEntry < len(p.recentIncomingBloomFilters) {
```

Comment:
Why is this condition? 
shouldn't all be deleted if `firstValidEntry == len(p.recentIncomingBloomFilters)` 
i.e. all the elements are less than `currentRound.SubSaturate(1)`

Reply:
hmm. yes.